src/main/java/io/burt/athena/result/s3/InputStreamResponseTransformer.java

```diff
@@ -98,26 +99,27 @@ private void maybeRequestMore(int currentSize) {
             if (newRequests < CHUNKS_REQUEST_LIMIT) {
                 if (newRequests * approximateChunkSize + currentSize < TARGET_BUFFER_SIZE) {
                     requests.addAndGet(10);
-                    subscription.request(10);
+                    subscription.get().ifPresent(s -> s.request(10L));
                 }
             }
         }
     }
 
     @Override
     public void onError(Throwable t) {
-        exceptionOccurred(t);
+        error = t;
+        onComplete();
     }
 
     @Override
     public void onComplete() {
         chunks.offer(END_MARKER);
-        complete.set(true);
+        subscription.getAndSet(Optional.empty()).ifPresent(Subscription::cancel);
```

Comment:
Well, the whole reason for the optional construct was basically because the three-line solution annoyed me, but you are probably right that it is too "clever" for its own good.

Reply:
I updated the implementation to skip Optional altogether. 

Also, I feel I should comment on:

> which I guess is not strictly thread safe, but if I remember correctly it would be safe in practice since the subscription never gets re-set.

Actually, with the addition in this PR, the subscription gets reset when the downloading is "resumed", so in theory there could be an issue.

However, the contract for the AsyncResponseTransformer class specifies that only one `onError` and `onComplete` will be called, and technically, in these cases you don't really need to cancel, as the subscription is effectively already cancelled by the producer at this point, but the contract also specifies that it should be safe to call `cancel` at this point.

Thus, the only situation where you actually "need" to cancel is if someone calls `#close` (before the other methods), and in theory this is a problem if close is called concurrently with a read retry subscription. However, the `InputStream` part of this class is already not thread-safe in many ways, so with only a single consumer, and the only way to initiate a resume from the resumers `#read` call, this should be safe.

I suppose in principle, it would make more sense to move the cancel call to the `#close` method.